### PR TITLE
Change script resources to use pipes rather than writing to temp files

### PR DESCRIPTION
**Pull request by phiggins**
Modifies the code for script resources to avoid writing temp files when not needed to avoid problems like the one described in #3563. 

There is some not-strictly-necessary cleanup on the Batch, PowershellScript, and WindowsScript providers that I did when I thought I was going to have to refactor the inheritance relationship between WindowsScript and the Script provider.

### CODE REVIEW

File: spec/unit/provider/powershell_script_spec.rb [hunk 1]
```diff
@@ -54,32 +32,12 @@ def set_user_defined_flag
...
-    end
-
-    it "sets default -ExecutionPolicy flag to 'Bypass'" do
-      expect(execution_policy_flag).to eq("Bypass")
-    end
-
-    it "sets user defined -ExecutionPolicy flag to 'RemoteSigned'" do
-      set_user_defined_flag
-      expect(execution_policy_flag).to eq("RemoteSigned")
-    end
```
Comment (phiggins) [inline spec/unit/provider/powershell_script_spec.rb:83]: This was testing the behavior of a property on the resource, so I moved the tests for that there.

File: spec/unit/provider/batch_spec.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      script_file_path = nil
+      provider.with_temp_script_file do
+        command = provider.command
+        script_file_path = provider.script_file_path
+      end
+
+      expect(command).to eq(%Q{"#{interpreter}" /f /c "#{script_file_path}"})
+    end
+  end
+end
```
Comment (phiggins) [inline spec/unit/provider/batch_spec.rb:130]: This was copied from the existing tests for the script provider and modified to work for batch.

File: spec/unit/provider/batch_spec.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  end
+
+  describe "#command" do
+    let(:basepath) { "C:\\Windows\\system32" }
+    let(:interpreter) { File.join(basepath, "cmd.exe") }
+
+    before do
+      allow(provider).to receive(:basepath).and_return(basepath)
+      provider.singleton_class.send(:public, :with_temp_script_file)
+      provider.singleton_class.send(:public, :script_file_path)
```
Comment (phiggins) [inline spec/unit/provider/batch_spec.rb:103]: This is kinda gross but I thought it was better than a bunch of stubbing and mocking.

File: lib/chef/provider/batch.rb [hunk 4]
```diff
@@ -27,11 +27,7 @@ class Batch < Chef::Provider::WindowsScript
       def command
         interpreter_path = Chef::Util::PathHelper.join(basepath, interpreter)
 
-        "\"#{interpreter_path}\" #{flags} \"#{script_file.path}\""
-      end
-
-      def flags
-        new_resource.flags.nil? ? "/c" : new_resource.flags + " /c"
+        "\"#{interpreter_path}\" #{new_resource.flags} /c \"#{script_file.path}\""
```
Comment (tas50) [inline lib/chef/provider/batch.rb]: Do the extra spaces we end up with here hurt us if flags is nil?

Comment (phiggins) [reply lib/chef/provider/batch.rb]: It doesn't seem to affect it.

**Comment by phiggins** [comment]
@tas50 brought up a concern that this wouldn't work on older versions of windows. In #9910 I tested that the underlying parts of this would work on various OSes but I used a windows 2019 VM. Borrowing the testing ideas from #9910 I was able to prove this does work in windows 2012r2.

Given this test script with the integers from 1 to 200,000 appended:

```
Petes-MBP:chef pete$ cat large_ruby_script_test.rb
require "digest"

data = DATA.binmode.read.encode(universal_newline: true)

=begin
Petes-MBP:chef pete$ seq 1 200000 | md5sum
0e10426a1d5bddffcef02f1345787128  -
=end
expected_md5 = "0e10426a1d5bddffcef02f1345787128"

puts "expected data to be 200,000 lines, got #{data.split.size}"

actual_md5 = Digest::MD5.hexdigest(data)
puts "expected digest:"
puts expected_md5
puts "got:"
puts actual_md5

# seq 1 200000 >> test_large_stdin.rb
__END__
```

This script is just over 1MB:

```
Petes-MBP:chef pete$ ls -l large_ruby_script_test.rb
-rw-r--r--  1 pete  staff  1289337 Jun  3 17:37 large_ruby_script_test.rb
```

I created a resource that reads that script:

```
Petes-MBP:chef pete$ cat test_ruby_resource_large_input.rb
ruby "test some ruby with a large input file" do
  code File.read("large_ruby_script_test.rb")
end
```

Running this in a windows 2012r2 VM gives the correct results:

```
PS C:\vagrant\chef> systeminfo | grep "^OS Name\|^OS Version"
OS Name:                   Microsoft Windows Server 2012 R2 Standard Evaluation
OS Version:                6.3.9600 N/A Build 9600
PS C:\vagrant\chef> ruby -Ilib .\chef-bin\bin\chef-apply .\test_ruby_resource_large_input.rb --log-level=debug
[2020-06-04T01:00:43+00:00] DEBUG: Reading products and relationships...
[2020-06-04T01:00:43+00:00] DEBUG: Successfully read products and relationships
[2020-06-04T01:00:43+00:00] DEBUG: Searching for the following licenses: ["infra-client", "inspec"]
[2020-06-04T01:00:43+00:00] DEBUG: Found license chef_infra_client at C:/Users/vagrant/.chef/accepted_licenses/chef_infr
a_client
[2020-06-04T01:00:43+00:00] DEBUG: Found license inspec at C:/Users/vagrant/.chef/accepted_licenses/inspec
[2020-06-04T01:00:43+00:00] DEBUG: Missing licenses remaining: []
[2020-06-04T01:00:43+00:00] DEBUG: All licenses present
[2020-06-04T01:00:43+00:00] DEBUG: Running Ohai with the following configuration: {:logger=>#<Mixlib::Log::Child:0x00000
0000985b408 @parent=Chef::Log, @metadata={:system=>"ohai", :version=>"16.1.1"}>}
[2020-06-04T01:00:43+00:00] DEBUG: Running Ohai with the following configuration: {:logger=>#<Mixlib::Log::Child:0x00000
0000985b408 @parent=Chef::Log, @metadata={:system=>"ohai", :version=>"16.1.1"}>}
[2020-06-04T01:00:48+00:00] DEBUG: Extracting run list from JSON attributes provided on command line
[2020-06-04T01:00:48+00:00] DEBUG: Applying attributes from json file
[2020-06-04T01:00:48+00:00] DEBUG: Platform is windows version 6.3.9600
[2020-06-04T01:00:48+00:00] INFO: Run List is []
[2020-06-04T01:00:48+00:00] INFO: Run List expands to []
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * ruby[test some ruby with a large input file] action run[2020-06-04T01:00:48+00:00] INFO: Processing ruby[test some r
uby with a large input file] action run ((chef-apply cookbook)::(chef-apply recipe) line 1)

    [execute] expected data to be 200,000 lines, got 200000
              expected digest:
              0e10426a1d5bddffcef02f1345787128
              got:
              0e10426a1d5bddffcef02f1345787128
[2020-06-04T01:00:48+00:00] INFO: ruby[test some ruby with a large input file] ran successfully
    - execute "ruby"  "C:/Users/vagrant/AppData/Local/Temp/1/chef-script20200604-1264-sxfn0g"
[2020-06-04T01:00:48+00:00] DEBUG: Exiting
```

**Pull request merged**